### ISSUE TITLE
Existing subscriptions on reinstallation

### ISSUE THREAD

**Issue by Kautenja** [mentrena/SyncKit#9]
hey i'm having a small issue with the update notifications feature. When setting this feature up on application launch, I attempt to subscribe the device for update notifications and check to make sure the device isn't subscribed already as such:

```swift
        var sync: QSCloudKitSynchronizer!
        ...
        // subscribe for update notifications
        if !sync.isSubscribedForUpdateNotifications() {
            sync.subscribeForUpdateNotifications { (error) in
                guard error == nil else {
                    NSLog(error!.localizedDescription)
                    return
                }
                NSLog("Remote iCloud updates live")
            }
        }
```

On first launch with a device that has no iCloud records yet logs `Remote iCloud updates live`. The issue I'm seeing is that `isSubscribedForUpdateNotifications()` seems to continue returning false on subsequent launches even when there is an existing subscription as indicated by the error that is logged: `Error saving record subscription with id 9BC150CC-ED7D-40B6-9CDD-E4E9D76EE68A to server: subscription is duplicate of '0E44F8CB-B3C5-411B-A056-0496190E6854'`

Despite the error, the device will still receive update notifications and everything still seems to work correctly so I'm not terribly concerned. Is this a bug with the `isSubscribedForUpdateNotifications()` getter or is my understanding incorrect with how to proceed here?

**Comment by Kautenja** [comment]
Update: deleting the subscription from iCloud manually and moving the block to setup update notifications _before_ the block to synchronize seems to have resolved the problem for now. i.e.

```swift
        // subscribe for update notifications
        if !sync!.isSubscribedForUpdateNotifications() {
            sync!.subscribeForUpdateNotifications { (error) in
                guard error == nil else {
                    NSLog(error!.localizedDescription)
                    return
                }
                NSLog("Remote Notifications live")
            }
        }
        // sync the data
        sync!.synchronize { (error) in
            guard error == nil else {
                NSLog(error!.localizedDescription)
                return
            }
            // post a notification so whatever view controller can update
            NotificationCenter.default.post(name: .cloudKitSync, object: self, userInfo: nil)
        }
```

 After inspecting the codebase it looks like when the subscription process throws an error in the callback here:

```objective-c
    [self.database saveSubscription:subscription completionHandler:^(CKSubscription * _Nullable subscription, NSError * _Nullable error) {
        if (!error) {
            [[NSUserDefaults standardUserDefaults] setObject:subscription.subscriptionID forKey:[self userDefaultsKeyForKey:QSSubscriptionIdentifierKey]];
        }
        
        callBlockIfNotNil(completion, error);
    }];
```

 the user default was never being set for the subscription putting me in an infinite loop until the subscription could be removed. Checking for this specific error and either suppressing it since it seems to not affect the functionality or handling it by removing the existing subscription and creating a new one would let the user default get set and prevent this error loop between launches. I'd open a PR, but I have no objective-c exp. easy fix though.

**Comment by Kautenja** [comment]
# TL;DR
* somehow my device got into an infinite error loop with subscriptions
* the issue isn't breaking
* changing the callback in `subscribeForChangesInRecordZoneWithCompletion` to check for an already existing subscription error and handling it accordingly will resolve future occurrences of this issue by allowing the user default for subscriptions to get reset allowing `isSubscribedForUpdateNotifications()` to return true

**Comment by Kautenja** [comment]
Update: this occurs when the application is uninstalled and reinstalled while a subscription is active. The subscription still exists in the cloud but the new installation has no knowledge of it. The error provides the id of the existing subscription so the way I see it there are 2 valid approaches here:

1. take the id of the existing subscription from the error and place it into user defaults, effectively restoring the connection to a valid subscription that still seems to send updates to the device. I can't see any reason why this would be problematic, but 2 is the safe route

2. delete the existing subscription to be safe and create a new one entirely. this guarantees that the subscription is valid and that the user default will be appropriately set. Since this error should only occur on reinstallation (or in other instances of user default corruption) i think it would be appropriate to handle it this more expensive way in the interest of protecting functionality over saving time

**Comment by mentrena** [comment]
The docs seem to suggest that creating the subscription once is enough --each device still will need to register for push notifications. So more like your first suggestion.

Should be fixed now on `0.4.6`, thanks for reporting!

**Comment by Kautenja** [comment]
can confirm the fix! thanks again lol. These two issues being resolved was literally like Christmas as they were of the last bugs I needed to resolve before I could launch v1.0 of a new app of mine. I'm hosting a public beta at http://www.3ringapp.com if you'd like to see your framework kicking ass and syncing data beautifully!
